wpilibc/src/main/native/cpp/LEDPattern.cpp

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#include "frc/LEDPattern.h"
+
+#include <algorithm>
+#include <cmath>
+#include <memory>
+#include <numbers>
+
+#include <wpi/MathExtras.h>
+#include <wpi/timestamp.h>
+
+using namespace frc;
+
+LEDPattern::LEDPattern(LEDPatternFn impl) : m_impl(std::move(impl)) {}
+
+void LEDPattern::ApplyTo(std::span<AddressableLED::LEDData> data,
+                         LEDWriterFn writer) const {
+  m_impl(data, writer);
+}
+
+void LEDPattern::ApplyTo(std::span<AddressableLED::LEDData> data) const {
+  ApplyTo(data, [&](int index, Color color) { data[index].SetLED(color); });
+}
+
+LEDPattern LEDPattern::Reversed() {
+  return LEDPattern{[self = *this](auto data, auto writer) {
+    self.ApplyTo(data, [&](int i, Color color) {
+      writer((data.size() - 1) - i, color);
+    });
+  }};
+}
+
+LEDPattern LEDPattern::OffsetBy(int offset) {
+  return LEDPattern{[=, self = *this](auto data, auto writer) {
+    self.ApplyTo(data, [&data, &writer, offset](int i, Color color) {
+      int shiftedIndex = floorMod(i + offset, data.size());
+      writer(shiftedIndex, color);
+    });
+  }};
+}
+
+LEDPattern LEDPattern::ScrollAtRelativeSpeed(units::hertz_t velocity) {
+  // velocity is in terms of LED lengths per second (1_hz= full cycle per
+  // second, 0.5_hz = half cycle per second, 2_hz = two cycles per second)
+  // Multiply by 1,000,000 to get microseconds
+  double periodMicros = 1e6 / velocity.value();
+
+  return LEDPattern{[=, self = *this](auto data, auto writer) {
+    auto bufLen = data.size();
+    auto now = wpi::Now();
+
+    // index should move by (bufLen) / (period)
+    double t =
+        (now % static_cast<uint64_t>(std::floor(periodMicros))) / periodMicros;
```

Comment:
UBSAN is complaining here: `runtime error: -2.56e+08 is outside the range of representable values of type 'unsigned long'` because it's negative.  Does this need to be int64_t, or does it need to be the absolute value?

Reply:
Yep, that's a copy-paste error from when I was converting C-style casts to `static_cast`